File: libs/power-log-parser/src/lib/parsers/healing-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEvent, GameEventProvider } from '../game-event';
import { Action, MetaData, Node, NodeType } from '../models';
import { MetaDataType } from '../enums';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

interface HealingInternal {
	SourceEntityId: number;
	SourceControllerId: number;
	TargetEntityId: number;
	TargetControllerId: number;
	Healing: number;
	Timestamp: string;
}

export class HealingParser implements ActionParser {
	readonly ParserName = 'HealingParser';

	private GameState: GameState;
	private Helper: StateFacade;

	constructor(parserState: ParserState, helper: StateFacade) {
		this.GameState = parserState.GameState;
		this.Helper = helper;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.Action &&
			this.hasHealingTag(node.Object as Action)
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const healingTags = action.Data.filter((d): d is MetaData => d instanceof MetaData).filter(
			(meta) => meta.Meta === (MetaDataType.HEALING as number),
		);
		const totalHealings: Record<string, Record<string, HealingInternal>> = {};
		for (const healingTag of healingTags) {
			for (const info of healingTag.MetaInfo) {
				const healingTarget = this.GameState.CurrentEntities.get(info.Entity)!;
				const targetEntityId = healingTarget.Id;
				const targetCardId = this.GameState.GetCardIdForEntity(healingTarget.Id);
				const targetControllerId = healingTarget.GetEffectiveController();
				const healingSource = this.getHealingSource(action);
				const sourceEntityId = healingSource.Id;
				const sourceCardId = this.GameState.GetCardIdForEntity(healingSource.Id);
				const sourceControllerId = healingSource.GetEffectiveController();

				let currentSourceHealings = totalHealings[sourceCardId!];
				if (!currentSourceHealings) {
					currentSourceHealings = {};
					totalHealings[sourceCardId!] = currentSourceHealings;
				}

				let currentTargetHealings = currentSourceHealings[targetCardId!];
				if (!currentTargetHealings) {
					currentTargetHealings = {
						SourceEntityId: sourceEntityId,
						SourceControllerId: sourceControllerId,
						TargetEntityId: targetEntityId,
						TargetControllerId: targetControllerId,
						Healing: 0,
						Timestamp: info.TimeStamp,
					};
					currentSourceHealings[targetCardId!] = currentTargetHealings;
				}
				currentTargetHealings.Healing = currentTargetHealings.Healing + healingTag.Data;
			}
		}

		const result: GameEventProvider[] = [];
		for (const healingSource of Object.keys(totalHealings)) {
			const targetHealings = totalHealings[healingSource];
			const firstTarget = Object.values(targetHealings)[0];
			const timestamp = firstTarget.Timestamp;
			result.push(
				GameEventProvider.Create(
					timestamp,
					'HEALING',
					() =>
						({
							Type: 'HEALING',
							Value: {
								SourceCardId: healingSource,
								SourceEntityId: Object.values(targetHealings)[0].SourceEntityId,
								SourceControllerId: Object.values(targetHealings)[0].SourceControllerId,
								Targets: targetHealings,
								LocalPlayer: this.Helper.LocalPlayer,
								OpponentPlayer: this.Helper.OpponentPlayer,
							},
						}) as GameEvent,
					true,
					node,
				),
			);
		}

		return result;
	}

	private hasHealingTag(action: Action): boolean {
		return action.Data.filter((d): d is MetaData => d instanceof MetaData).some(
			(meta) => meta.Meta === (MetaDataType.HEALING as number),
		);
	}

	private getHealingSource(action: Action): any {
		return this.GameState.CurrentEntities.get(action.Entity)!;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/HealingParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData.Meta;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class HealingParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade Helper { get; set; }

        public HealingParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.Helper = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(Parser.ReplayData.GameActions.Action)
                && HashHealingTag(node.Object as Parser.ReplayData.GameActions.Action);
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Parser.ReplayData.GameActions.Action;
            var healingTags = action.Data
                .Where((d) => d.GetType() == typeof(MetaData))
                .Select((meta) => meta as MetaData)
                .Where((meta) => meta.Meta == (int)MetaDataType.HEALING);
            Dictionary<string, Dictionary<string, HealingInternal>> totalHealings = new Dictionary<string, Dictionary<string, HealingInternal>>();
            foreach (var healingTag in healingTags)
            {
                foreach (var info in healingTag.MetaInfo)
                {
                    // If source or target are player entities, they don't have any 
                    // attached cardId
                    var healingTarget = GameState.CurrentEntities[info.Entity];
                    var targetEntityId = healingTarget.Id;
                    var targetCardId = GameState.GetCardIdForEntity(healingTarget.Id);
                    var targetControllerId = healingTarget.GetEffectiveController();
                    var healingSource = GetHealingSource(healingTarget, action, healingTag);
                    var sourceEntityId = healingSource.Id;
                    var sourceCardId = GameState.GetCardIdForEntity(healingSource.Id);
                    var sourceControllerId = healingSource.GetEffectiveController();
                    Dictionary<string, HealingInternal> currentSourceHealings = null;
                    if (totalHealings.ContainsKey(sourceCardId))
                    {
                        currentSourceHealings = totalHealings[sourceCardId];
                    }
                    else
                    {
                        currentSourceHealings = new Dictionary<string, HealingInternal>();
                        totalHealings[sourceCardId] = currentSourceHealings;
                    }
                    HealingInternal currentTargetHealings = null;
                    if (currentSourceHealings.ContainsKey(targetCardId))
                    {
                        currentTargetHealings = currentSourceHealings[targetCardId];
                    }
                    else
                    {
                        currentTargetHealings = new HealingInternal
                        {
                            SourceEntityId = sourceEntityId,
                            SourceControllerId = sourceControllerId,
                            TargetEntityId = targetEntityId,
                            TargetControllerId = targetControllerId,
                            Healing = 0,
                            Timestamp = info.TimeStamp,
                        };
                        currentSourceHealings[targetCardId] = currentTargetHealings;
                    }
                    currentTargetHealings.Healing = currentTargetHealings.Healing + healingTag.Data;
                }
            }

            List<GameEventProvider> result = new List<GameEventProvider>();
            // Now send one event per source
            foreach (var healingSource in totalHealings.Keys)
            {
                var targetHealings = totalHealings[healingSource];
                var timestamp = totalHealings[healingSource].First().Value.Timestamp;
                result.Add(GameEventProvider.Create(
                    timestamp,
                    "HEALING",
                    // The structure of this event is too specific to be added to the generic GameEvent.CreateProvider() method
                    () => new GameEvent
                    {
                        Type = "HEALING",
                        Value = new
                        {
                            SourceCardId = healingSource,
                            SourceEntityId = totalHealings[healingSource].First().Value.SourceEntityId,
                            SourceControllerId = totalHealings[healingSource].First().Value.SourceControllerId,
                            Targets = totalHealings[healingSource],
                            LocalPlayer = Helper.LocalPlayer,
                            OpponentPlayer = Helper.OpponentPlayer,
                        }
                    },
                    true,
                    node));
            }

            return result;
        }

        private bool HashHealingTag(Parser.ReplayData.GameActions.Action action)
        {
            var data = action.Data;
            var numberOfHealingTags = data
                .Where((d) => d.GetType() == typeof(MetaData))
                .Where((meta) => (meta as MetaData).Meta == (int)MetaDataType.HEALING)
                .Count();
            return numberOfHealingTags > 0;
        }

        private FullEntity GetHealingSource(FullEntity target, Parser.ReplayData.GameActions.Action action, MetaData meta)
        {
            var actionSource = GameState.CurrentEntities[action.Entity];
            return actionSource;
        }

        private class HealingInternal
        {
            public int SourceEntityId;
            public int SourceControllerId;
            public int TargetEntityId;
            public int TargetControllerId;
            public int Healing;
            public DateTime Timestamp;
        }
    }
}
